### PR TITLE
Walletconnect support

### PR DESCRIPTION
**Pull request by makoto**

### CODE REVIEW

**Comment by makoto** [comment]
For some reason, it's making over 100 JSONRPC request to Ethereum node when user lands on the page as a read only mode (as compared to current page which has less than 30 API request).

Screenshot 2020-11-20 at 20 53 45

When I did some digging, it looks like many of API call to ENS such as `getOwner` seems not cached, so `ENS.owner(namehash)` of the same name seems called over 50 times.

```
  async getOwner(name) {
    const namehash = getNamehash(name)
    const owner = await this.ENS.owner(namehash)
    return owner
  }
```

This does not seem related to Apollo Client's fetch policy as it occurs even when I comment out `fetchPolicy: 'network-only',` config at `apolloClient.js`

This seems to happen when the user lands on the page for the first time. When user connect/disconnect wallet, it has about 30 JSON RPC call which is similar to what it is now.

I also tried to add caching like this but this will cause problem when users try to update resolver, set record, etc as cache is not going to be cleared.


```
  async getResolver(name) {
    if(!this.getResolverCache[name]){
      const namehash = getNamehash(name)
      this.getResolverCache[name] = this.ENS.resolver(namehash)  
    }
    return this.getResolverCache[name]
  }
```

I can do extra things by adding local storage to remember user's choice, but it would be nice if it doesn't call JSON RPC over 100 times to begin with (as this may be causing slow loading time on our CI env and tests becoming very flakey)

File: src/routes/Favourites.js [hunk 1]
```diff
@@ -166,32 +170,36 @@ function Favourites() {
...
-                <Query
-                  query={GET_OWNER}
-                  variables={{ name: domain.name }}
-                  key={domain.name}
-                >
-                  {({ loading, error, data }) => {
-                    if (error)
+              {data &&
+                data.subDomainFavourites &&
+                data.subDomainFavourites.map(domain => (
```
Comment (jefflau) [inline src/routes/Favourites.js:175]: optional chaining

File: src/routes/Home.js [hunk 2]
```diff
@@ -18,6 +20,7 @@ import HowToUseDefault from '../components/HowToUse/HowToUse'
 import Alice from '../components/HomePage/Alice'
 import ENSLogo from '../components/HomePage/images/ENSLogo.svg'
 import { abougPageURL } from '../utils/utils'
```
Comment (jefflau) [inline src/routes/Home.js:22]: Just spotted this, unrelated but do you mind changing aboug to about lol

File: src/routes/Home.js [hunk 3]
```diff
@@ -261,9 +269,28 @@ const PermanentRegistrarLogo = styled(motion.h1)`
...
+  const {
+    data: { getReverseRecord } = {},
+    loading: reverseRecordLoading
+  } = useQuery(GET_REVERSE_RECORD, {
+    variables: {
+      address
+    }
+  })
+  const displayName =
+    getReverseRecord && getReverseRecord.name
```
Comment (jefflau) [inline src/routes/Home.js:291]: optional chaining

File: src/routes/Home.js [hunk 4]
```diff
@@ -261,9 +269,28 @@ const PermanentRegistrarLogo = styled(motion.h1)`
...
+      ? getReverseRecord.name
+      : address && `${address.slice(0, 10)}...`
```
Comment (jefflau) [inline src/routes/Home.js:293]: optional chaining (if u can), not sure if u can do that with spread...

File: src/utils/utils.js [hunk 5]
```diff
@@ -182,3 +186,50 @@ export const abougPageURL = () => {
...
+        : {}
+      window.localStorage.setItem(
+        'labels',
+        JSON.stringify({
+          ...labels,
+          ...JSON.parse(process.env.REACT_APP_LABELS)
+        })
+      )
+    } else {
+      console.log('***handleNetworkChange1')
```
Comment (jefflau) [inline src/utils/utils.js:213]: remove

File: src/utils/utils.js [hunk 6]
```diff
@@ -182,3 +186,50 @@ export const abougPageURL = () => {
...
+      await setup({
+        reloadOnAccountsChange: false,
+        enforceReadOnly: true,
+        enforceReload: true
+      })
+      console.log('***handleNetworkChange2')
```
Comment (jefflau) [inline src/utils/utils.js:219]: remove

File: src/utils/utils.js [hunk 7]
```diff
@@ -182,3 +186,50 @@ export const abougPageURL = () => {
...
+    }
+    networkId = await getNetworkId()
+    console.log('***handleNetworkChange3', { networkId })
```
Comment (jefflau) [inline src/utils/utils.js:222]: remove

File: src/utils/utils.js [hunk 8]
```diff
@@ -182,3 +186,50 @@ export const abougPageURL = () => {
...
+    client = await setupClient(networkId)
+  } catch (e) {
+    networkId = networkId || 1 // Readonly to Mainnet
+    console.log('***handleNetworkChange4', { networkId })
```
Comment (jefflau) [inline src/utils/utils.js:226]: remove

File: src/utils/utils.js [hunk 9]
```diff
@@ -182,3 +186,50 @@ export const abougPageURL = () => {
...
+    client = await setupClient()
+    await client.mutate({
+      mutation: SET_ERROR,
+      variables: { message: e.message }
+    })
+  }
+  console.log('***handleNetworkChange5', { client, networkId })
```
Comment (jefflau) [inline src/utils/utils.js:233]: remove

File: src/components/NetworkInformation/NetworkInformation.js [hunk 10]
```diff
@@ -140,8 +173,8 @@ function NetworkInformation() {
   }
 
   return (
-    <NetworkInformationContainer hasAccount={accounts.length > 0}>
-      {accounts.length > 0 ? (
+    <NetworkInformationContainer hasAccount={accounts && accounts.length > 0}>
+      {accounts && accounts.length > 0 ? (
```
[the remark this answers is not in the document]
Comment (makoto) [reply src/components/NetworkInformation/NetworkInformation.js:177]: It does not seem to be supported on Opera & Android which has Ethereum wallet support (and also we don't know if some embeded wallet such as ImToken, Status etc would support

[the remark this answers is not in the document]
Comment (makoto) [reply src/components/NetworkInformation/NetworkInformation.js:177]: Also IE does not support which is big. Let's not use optional chaining.

File: src/components/DomainItem/DomainItem.js [hunk 11]
```diff
@@ -169,7 +169,10 @@ const Domain = ({ domain, isSubDomain, className, isFavourite, loading }) => {
           domain.owner &&
           parseInt(domain.owner, 16) !== 0
         ) {
-          isOwner = domain.owner.toLowerCase() === account.toLowerCase()
+          isOwner =
+            account &&
+            domain.owner &&
+            domain.owner.toLowerCase() === account.toLowerCase()
```
[the remark this answers is not in the document]
Comment (makoto) [reply src/components/DomainItem/DomainItem.js:175]: optional chaining is not supported on IE and opera.

File: src/components/SearchName/DomainInfo.js [hunk 12]
```diff
@@ -32,6 +32,7 @@ const DomainInfoContainer = ({ searchTerm }) => {
                 domain={singleName}
                 isFavourite={
                   singleName &&
+                  favourites &&
```
[the remark this answers is not in the document]
Comment (makoto) [reply src/components/SearchName/DomainInfo.js:35]: Same. optional chaining is not supported on IE and opera.

File: src/components/SearchName/Search.js [hunk 13]
```diff
@@ -80,17 +79,22 @@ function Search({ history, className, style }) {
   const handleParse = e => {
     setInputValue(e.target.value)
   }
-
+  const hasSearch = inputValue && inputValue.length > 0
```
[the remark this answers is not in the document]
Comment (makoto) [reply src/components/SearchName/Search.js:82]: Same. optional chaining is not supported on IE and opera.

File: src/components/SingleName/ResolverAndRecords/Coins/Coins.js [hunk 14]
```diff
@@ -20,9 +20,12 @@ export default function Coins(props) {
   return (
     <KeyValueRecord
       {...props}
-      records={props.updatedRecords.coins.sort(record =>
-        record.key === 'ETH' ? -1 : 1
-      )}
+      records={
+        props.updatedRecords.coins &&
+        props.updatedRecords.coins.sort(record =>
```
[the remark this answers is not in the document]
Comment (makoto) [reply src/components/SingleName/ResolverAndRecords/Coins/Coins.js:25]: Same. optional chaining is not supported on IE and opera.

File: src/components/SingleName/ResolverAndRecords/KeyValueRecord/KeyValueRecord.js [hunk 15]
```diff
@@ -246,37 +246,38 @@ function Records({
...
-            // Value empty
-            (value === emptyAddress || value === '') &&
-            // Value has not been changed
-            !changedRecords[recordType].find(record => record.key === key) &&
-            // Value is not a placeholder
-            !placeholderRecords.includes(key)
-          ) {
-            return null
-          }
+        {records &&
```
[the remark this answers is not in the document]
Comment (makoto) [reply src/components/SingleName/ResolverAndRecords/KeyValueRecord/KeyValueRecord.js:249]: Same. optional chaining is not supported on IE and opera.

File: src/globalState.js [hunk 16]
```diff
@@ -34,12 +36,17 @@ export class GlobalStateProvider extends Component {
...
+  }
+
   toggleModal = modal => {
-    this.setState(
-      state =>
-        state.currentModal && state.currentModal.name === modal.name
-          ? { currentModal: null }
-          : { currentModal: modal }
+    this.setState(state =>
+      state.currentModal && state.currentModal.name === modal.name
```
[the remark this answers is not in the document]
Comment (makoto) [reply src/globalState.js:47]: Same. optional chaining is not supported on IE and opera.

File: src/routes/Favourites.js (hunk 1 above)
Comment (makoto) [reply src/routes/Favourites.js:175]: Same. optional chaining is not supported on IE and opera.

File: src/components/NetworkInformation/NetworkInformation.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (jefflau) [reply src/components/NetworkInformation/NetworkInformation.js:177]: It's supported by babel, so we're fine

**Pull request merged**